File: mobile/app/src/main/java/cit/edu/garol/campus/features/claimrequest/api/ClaimRequestApi.kt
package cit.edu.garol.campus.features.claimrequest.api

import cit.edu.garol.campus.features.claimrequest.model.ClaimRequest
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Response
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Part
import retrofit2.http.Path
import retrofit2.http.Query

interface ClaimRequestApi {

    /*
     * ==========================================================
     * ADMIN - GET ALL CLAIM REQUESTS
     * ==========================================================
     */

    @GET("api/claim-requests")
    suspend fun getAllClaimRequests():
            Response<List<ClaimRequest>>

    /*
     * ==========================================================
     * USER - GET OWN CLAIM REQUESTS
     * ==========================================================
     */

    @GET("api/claim-requests/user/{claimantId}")
    suspend fun getUserClaimRequests(
        @Path("claimantId")
        claimantId: String
    ): Response<List<ClaimRequest>>

    /*
     * ==========================================================
     * ADMIN - APPROVE CLAIM REQUEST
     * ==========================================================
     */

    @PUT("api/claim-requests/{id}/approve")
    suspend fun approveClaimRequest(
        @Path("id")
        claimRequestId: Int
    ): Response<String>

    /*
     * ==========================================================
     * ADMIN - REJECT CLAIM REQUEST
     * ==========================================================
     */

    @PUT("api/claim-requests/{id}/reject")
    suspend fun rejectClaimRequest(
        @Path("id")
        claimRequestId: Int
    ): Response<String>

    /*
     * ==========================================================
     * ADMIN - UPDATE CLAIM STATUS
     * ==========================================================
     */

    @PUT("api/claim-requests/{id}/status")
    suspend fun updateClaimStatus(
        @Path("id")
        claimRequestId: Int,

        @Query("status")
        status: String
    ): Response<String>

    /*
 * ==========================================================
 * USER - SUBMIT CLAIM REQUEST
 * ==========================================================
 */

    @Multipart
    @POST("api/claim-requests")
    suspend fun submitClaimRequest(

        @Part("itemReportId")
        itemReportId: RequestBody,

        @Part("claimantId")
        claimantId: RequestBody,

        @Part("claimantName")
        claimantName: RequestBody,

        @Part("claimantEmail")
        claimantEmail: RequestBody,

        @Part("claimantPhone")
        claimantPhone: RequestBody,

        @Part("itemDescription")
        itemDescription: RequestBody,

        @Part("additionalInformation")
        additionalInformation: RequestBody,

        @Part
        proofImage: MultipartBody.Part?

    ): Response<ClaimRequest>

}

File: web/src/request-claim/claim_request_api.ts
import API_BASE_URL from "../api/api_config";

const CLAIM_REQUEST_API_URL =
  `${API_BASE_URL}/api/claim-requests`;
/*
 * Submit a claim request.
 */
export async function submitClaimRequest(
  formData: FormData
) {
  const response = await fetch(
    CLAIM_REQUEST_API_URL,
    {
      method: "POST",
      body: formData,
    }
  );

  if (!response.ok) {
    const message = await response.text();

    throw new Error(
      message || "Failed to submit claim request."
    );
  }

  return response.json();
}

/*
 * Admin - View all claim requests.
 */
export async function getAllClaimRequests() {
  const response = await fetch(
    CLAIM_REQUEST_API_URL
  );

  if (!response.ok) {
    const message = await response.text();

    throw new Error(
      message || "Failed to fetch claim requests."
    );
  }

  return response.json();
}

/*
 * User - View their own claim requests.
 */
export async function getUserClaimRequests(
  claimantId: string
) {
  const response = await fetch(
    `${CLAIM_REQUEST_API_URL}/user/${encodeURIComponent(
      claimantId
    )}`
  );

  if (!response.ok) {
    const message = await response.text();

    throw new Error(
      message || "Failed to fetch your claim requests."
    );
  }

  return response.json();
}

/*
 * Admin - Update claim request status.
 */
export async function updateClaimRequestStatus(
  id: number,
  status: string
) {
  const response = await fetch(
    `${CLAIM_REQUEST_API_URL}/${id}/status?status=${encodeURIComponent(
      status
    )}`,
    {
      method: "PUT",
    }
  );

  if (!response.ok) {
    const message = await response.text();

    throw new Error(
      message || "Failed to update claim request."
    );
  }

  return response.json();
}